src/range.hpp

```diff
@@ -0,0 +1,26 @@
+/*
+	Copyright (C) 2008 - 2023
+	Part of the Battle for Wesnoth Project https://www.wesnoth.org/
+
+	This program is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY.
+
+	See the COPYING file for more details.
+*/
+
+#pragma once
+
+#include "enum_base.hpp"
+
+struct range_defines
+{
+	static constexpr const char* const melee = "melee";
+	static constexpr const char* const ranged = "ranged";
+
+	ENUM_AND_ARRAY(melee, ranged)
```

Comment:
I suggest removing it from the unit editor as well and making it a combobox with melee and ranged prefilled as options.

Reply:
I still don't get why an editable combobox is more useful than a simple menu for selecting weapon ranges. While it should have some way to give a unit attacks with `type=secret`, the editable combobox seems likely to result in typos such as `mele` more often than actually wanted variations.

Reply:
If that's the case, then we shouldn't allow people to write their WML/lua either - just imagine how many typos that's caused!

More seriously, I think it's not our problem. If someone typos a custom value for a field that can have custom values defined, then that's really their own fault.

Reply:
I'm in favor of a UI that allows custom values, my point is that IMO we can make a better UI, including the custom value support, with the existing components and the two other new components from this PR than with the editable combobox. I think it will be as easy or easier, because the combobox on offer here already needs the code using it to handle a lot of the coordination between editable part and list.

My question to @CelticMinstrel is, why is a combobox with the API that's in this PR preferable for 1.18? While we could change the C++ API of it for 1.19, if we include it in 1.18 then we'll be stuck with supporting its GUI2 WML API.

Reply:
> we'll be stuck with supporting its GUI2 WML API.

What would you even imagine changing about the GUI2 WML API presented here? I can't see anything that's obviously controversial, except maybe the "history" key (does a combobox need history? not sure).

Reply:
Removed range.hpp.

Reply:
(Although the line that this was opened on was resolved, this seems a good thread for discussing the combobox.)

Mock-up of an alternative to the combobox, at least for 1.18. Selecting a value (other than Custom) in the `menu_button` would overwrite the `text_box`, typing in the `text_box` would automatically select a value in the `menu_button`. That allows custom values, while still making typos obvious.
![image](https://github.com/wesnoth/wesnoth/assets/101462/915fd404-d080-4ae5-9921-7f0b587514c0)

> would you even imagine changing about the GUI2 WML API presented here?

You're asking me to quickly do a lot of the thought that would come up during a full design and testing cycle, and you haven't shown why a combobox is needed. If we include it, we have the future maintenance burden, even if it ends up unused. We lose the option to not have a combobox, at least not without a deprecation cycle.

I'd add a boolean of whether the text box is a search term (allowing only the existing values to be chosen) or a text field that allows other value.

I'd add a boolean of whether there's visual feedback when the text box contains a value that isn't in the list.

We could add those booleans later, but that comes back to the question of why it's useful to rush the current spec into 1.18.

Reply:
Let's see what @CelticMinstrel has to say, but from me :
1. I can remove the Combobox from here and put it in another PR intended for 1.19.
2. Since I'm not supporting custom alignments in the Unit Editor ATM, so I can just keep the alignment box as a normal Menu Button, and add support for custom values later (with corresponding changes to the UI).
@stevecotton Does this sound good to you?

Reply:
BTW we can get rid of those "copy text" buttons I see almost everywhere in wesnoth if the editable property is used.

Reply:
> Mock-up of an alternative to the combobox, at least for 1.18. Selecting a value (other than Custom) in the `menu_button` would overwrite the `text_box`, typing in the `text_box` would automatically select a value in the `menu_button`. That allows custom values, while still making typos obvious.

Your mockup is _literally_ a combobox, just a much more bulky implementation thereof. If you really like that layout, I bet it could just be one of the combobox's definitions, but I'm not sure how that could possibly be better than the actual combobox that has been implemented. It takes up twice as much space for no added value.

> you haven't shown why a combobox is needed

I'm pretty sure this is not true. Comboboxes are needed for places where there are a set of common values but you can also use custom values. They're also useful for places where you'll likely want to use the same values frequently – in this case the combobox lists all values that are in use, but you can also type in a new one. That second use-case might not arise in Wesnoth, but the first one definitely does.

> I'd add a boolean of whether the text box is a search term (allowing only the existing values to be chosen) or a text field that allows other value.

Um… what? If only the existing values can be chosen, why would you use a combobox? Just use a regular menu button.

If you're looking for a way to search for a value in a very long dropdown menu, I think I'd prefer putting a search field _inside_ the menu (perhaps it's invisible until you start typing).

> I'd add a boolean of whether there's visual feedback when the text box contains a value that isn't in the list.

I don't think visual feedback of whether the current value is in the list is a high priority for a combobox. It's not useless though, so it could be nice to have. But at minimum I would say the dropdown menu should preselect the current value, which means that if you have a custom value, no value is preselected in the dropdown menu.

> We could add those booleans later, but that comes back to the question of why it's useful to rush the current spec into 1.18.

I don't see the harm in merging this into 1.18. We can certainly add the boolean for visual feedback later even if the combobox already exists. But I also don't mind _that_ much if it's partially or wholly delayed to 1.19.